polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -534,174 +562,96 @@ Value SCFParallelGuardBuilder::createGuardExpr() const {
-    std::get<0>(it).replaceUsesWithIf(std::get<1>(it), [&](OpOperand &op) {
-      Block *useBlock = op.getOwner()->getBlock();
-      return useBlock != ifOp.getThenBlock();
-    });
+  SmallVector<AffineExpr, 2> exprs;
+  getConstraints(exprs, dims, symbols);
+  SmallVector<bool, 2> eqflags(exprs.size(), false);
+
+  return IntegerSet::get(
+      /*dim*/ lbMap.getNumDims() + ubMap.getNumDims(),
+      /*symbols*/ lbMap.getNumSymbols() + ubMap.getNumSymbols(), exprs,
+      eqflags);
 }
 
 //===----------------------------------------------------------------------===//
-// AffineParallelLoopGuardBuilder
+// AffineForLoopGuardBuilder
 //===----------------------------------------------------------------------===//
 
-void AffineParallelGuardBuilder::guardLoop() const { guard(loop); }
-
-IntegerSet AffineParallelGuardBuilder::createGuardExpr() const {
-  SmallVector<AffineExpr, 2> exprs;
-  SmallVector<bool, 2> eqflags;
+void AffineForGuardBuilder::getConstraints(SmallVectorImpl<AffineExpr> &exprs,
+                                           ArrayRef<AffineExpr> dims,
+                                           ArrayRef<AffineExpr> symbols) const {
+  for (AffineExpr ub : loop.getUpperBoundMap().getResults()) {
+    ub = ub.replaceDimsAndSymbols(dims, symbols);
+    for (AffineExpr lb : loop.getLowerBoundMap().getResults()) {
+      // Bound is whether this expr >= 0, which since we want ub > lb, we
+      // rewrite as follows.
+      exprs.push_back(ub - lb - 1);
+    }
+  }
+}
 
-  const AffineMap lbMap = loop.getLowerBoundsMap(),
-                  ubMap = loop.getUpperBoundsMap();
+//===----------------------------------------------------------------------===//
+// AffineParallelLoopGuardBuilder
+//===----------------------------------------------------------------------===//
 
-  for (auto step : llvm::enumerate(loop.getSteps())) {
+void AffineParallelGuardBuilder::getConstraints(
+    SmallVectorImpl<AffineExpr> &exprs, ArrayRef<AffineExpr> dims,
+    ArrayRef<AffineExpr> symbols) const {
+  for (auto step : llvm::enumerate(loop.getSteps()))
     for (AffineExpr ub : loop.getUpperBoundMap(step.index()).getResults()) {
-      SmallVector<AffineExpr, 4> symbols;
-      for (unsigned idx = 0; idx < ubMap.getNumSymbols(); ++idx)
-        symbols.push_back(getAffineSymbolExpr(idx + lbMap.getNumSymbols(),
-                                              loop.getContext()));
-
-      SmallVector<AffineExpr, 4> dims;
-      for (unsigned idx = 0; idx < ubMap.getNumDims(); ++idx)
-        dims.push_back(
-            getAffineDimExpr(idx + lbMap.getNumDims(), loop.getContext()));
-
       ub = ub.replaceDimsAndSymbols(dims, symbols);
-
       for (AffineExpr lb : loop.getLowerBoundMap(step.index()).getResults()) {
         // Bound is whether this expr >= 0, which since we want ub > lb, we
         // rewrite as follows.
-        exprs.push_back(ub - lb - step.value());
-        eqflags.push_back(false);
+        exprs.push_back(ub - lb - 1);
```

Comment:
This is the only functional change, the loop guard should not depend on loop step.

Reply:
Can you add a test with step larger than 1 to cover this?